src/analysis/normed_space/normed_group_hom.lean

```diff
@@ -0,0 +1,358 @@
+
+@[simp] lemma map_zero : f 0 = 0 := f.to_add_monoid_hom.map_zero
+
+@[simp] lemma map_add (x y) : f (x + y) = f x + f y := f.to_add_monoid_hom.map_add _ _
+
+@[simp] lemma map_sum {ι : Type*} (v : ι → V₁) (s : finset ι) :
+  f (∑ i in s, v i) = ∑ i in s, f (v i) :=
+f.to_add_monoid_hom.map_sum _ _
+
+@[simp] lemma map_sub (x y) : f (x - y) = f x - f y := f.to_add_monoid_hom.map_sub _ _
+
+@[simp] lemma map_neg (x) : f (-x) = -(f x) := f.to_add_monoid_hom.map_neg _
+
+/-- Make a normed group hom from a group hom and a norm bound. -/
+def mk' (f : V₁ →+ V₂) (C : ℝ≥0) (hC : ∀ x, ∥f x∥ ≤ C * ∥x∥) : normed_group_hom V₁ V₂ :=
+{ bound' := ⟨C, hC⟩ .. f}
+
+@[simp] lemma coe_mk' (f : V₁ →+ V₂) (C) (hC) : ⇑(mk' f C hC) = f := rfl
+
+/-- Predicate asserting a norm bound on a normed group hom. -/
+def bound_by (f : normed_group_hom V₁ V₂) (C : ℝ≥0) : Prop := ∀ x, ∥f x∥ ≤ C * ∥x∥
+
+lemma mk'_bound_by (f : V₁ →+ V₂) (C) (hC) : (mk' f C hC).bound_by C := hC
+
+lemma bound : ∃ C, 0 < C ∧ f.bound_by C :=
+begin
+  obtain ⟨C, hC⟩ := f.bound',
+  let C' : ℝ≥0 := ⟨max C 1, le_max_right_of_le zero_le_one⟩,
+  use C',
+  simp only [C', ← nnreal.coe_lt_coe, subtype.coe_mk, nnreal.coe_zero,
+    lt_max_iff, zero_lt_one, or_true, true_and],
+  intro v,
+  calc ∥f v∥
+      ≤ C * ∥v∥ : hC v
+  ... ≤ max C 1 * ∥v∥ : mul_le_mul (le_max_left _ _) le_rfl (norm_nonneg _) _,
+  exact zero_le_one.trans (le_max_right _ _)
+end
+
+lemma lipschitz_of_bound_by (C : ℝ≥0) (h : f.bound_by C) :
+  lipschitz_with (nnreal.of_real C) f :=
+lipschitz_with.of_dist_le' $ λ x y, by simpa only [dist_eq_norm, f.map_sub] using h (x - y)
+
+theorem antilipschitz_of_bound_by {K : ℝ≥0} (h : ∀ x, ∥x∥ ≤ K * ∥f x∥) :
+  antilipschitz_with K f :=
+antilipschitz_with.of_le_mul_dist $
+λ x y, by simpa only [dist_eq_norm, f.map_sub] using h (x - y)
+
+protected lemma uniform_continuous (f : normed_group_hom V₁ V₂) :
+  uniform_continuous f :=
+begin
+  obtain ⟨C, C_pos, hC⟩ := f.bound,
+  exact (lipschitz_of_bound_by f C hC).uniform_continuous
+end
+
+@[continuity]
+protected lemma continuous (f : normed_group_hom V₁ V₂) : continuous f :=
+f.uniform_continuous.continuous
+
+variables {f g}
+
+instance : has_zero (normed_group_hom V₁ V₂) :=
+⟨{ bound' := ⟨0, λ v, show ∥(0 : V₂)∥ ≤ 0 * _, by rw [norm_zero, zero_mul]⟩,
+   .. (0 : V₁ →+ V₂) }⟩
+
+instance : inhabited (normed_group_hom V₁ V₂) := ⟨0⟩
+
+lemma coe_inj ⦃f g : normed_group_hom V₁ V₂⦄ (h : (f : V₁ → V₂) = g) : f = g :=
+by cases f; cases g; cases h; refl
+
+/-- The identity as a continuous normed group hom. -/
+@[simps]
+def id : normed_group_hom V V :=
+{ bound' := ⟨1, λ v, show ∥v∥ ≤ 1 * ∥v∥, by rw [one_mul]⟩,
+  .. add_monoid_hom.id V }
+
+/-- The composition of continuous normed group homs. -/
+@[simps]
+def comp (g : normed_group_hom V₂ V₃) (f : normed_group_hom V₁ V₂) :
+  normed_group_hom V₁ V₃ :=
+{ bound' :=
+  begin
+    obtain ⟨Cf, Cf_pos, hf⟩ := f.bound,
+    obtain ⟨Cg, Cg_pos, hg⟩ := g.bound,
```

Comment:
All these `obtain` proofs seem a little bit clumsy. Why not define the norm on `normed_group_hom V V'` earlier in this file (possibly by copying/factoring out parts  of `normed_space/operator_norm.lean`) and then use the norm of `f` and `g` together with their basic properties?

Reply:
I thought for a bit, but I don't really see how you want to golf these.